### PR TITLE
Improve error messages

### PR DESCRIPTION
**Pull request by evansd**
This is a follow on from #1873 and attempts to address many of the issues noted in #505.

As part of this work I discovered that there's an alternative ISO date format (`YYYYMMDD` without hyphens) and so I've updated the parser to only accept the hyphenated version to avoid any possibility of confusion.

### CODE REVIEW

File: tests/unit/test_query_language.py [hunk 1]
```diff
@@ -74,6 +74,17 @@ class events(EventFrame):
 events_schema = TableSchema(event_date=Column(date), f=Column(float))
 
 
+def assert_not_chained_exception(excinfo):
+    # Including chained exception details in the traceback is the default Python
+    # behaviour but we often want to hide internal details from the user where these are
+    # not helpful
+    traceback_str = "\n".join(traceback.format_exception(excinfo.value))
```
Comment (bloodearnest) [inline tests/unit/test_query_language.py:82]: How about `assert excinfo.value.__context__ is None` instead or as well?

**Review by bloodearnest — APPROVED** [review]
This makes some sense to be at a high level, although I don't really understand the details of some of the specific errors.

File: ehrql/query_language.py [hunk 2]
```diff
@@ -30,12 +30,17 @@
 REGISTERED_TYPES = {}
 
 
-class InvalidOperationError(Exception):
+class LanguageError(Exception):
```
Comment (inglesp) [inline ehrql/query_language.py:33]: Could we call this `EhrqlError`?

**Review by inglesp — APPROVED** [review]

File: ehrql/query_language.py (hunk 2 above)
Comment (evansd) [reply ehrql/query_language.py:33]: Yeah, I considered that but got put off by the question of how to capitalise it! I think it's the more natural name though, if we can get over my hang-ups.

**Pull request merged**